### ISSUE TITLE
Finding references is very slow

### ISSUE THREAD

**Issue by danieltodor** [microsoft/pylance-release#2109]
## Environment data

-   Language Server version: Pylance: 2021.11.2 / Python extension: 2021.11.1422169775
-   OS and version: Archlinux (latest)
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9.7 64bit

## Expected behaviour

Using the commands `Peek References` or `Peek Definition` is expected to show references or definition (almost) immediately after it was used more than once.

## Actual behaviour

It regenerates the tags each time i use these commands. Looks like they are not stored in memory, so for me it takes 5-10 second each time i use these commands.
In the logs, it looks like it stores the tags, but still saying 'Finding References' on the bottom section of the window, and takes 5-10 seconds to show them.

I tested this on my other machine, and it works perfectly on it. The first time takes 5-10 seconds, then it shows the peek panel immediately every time. The only difference i see is that it is a Windows 10 machine with python version 3.9.5 64 bit. It has the same settings.json, same python and pylance extension versions, same project files, same VSCode version. The generated logs are exactly the same.

## Logs

```
Log from "Python":
Experiment 'pythonaa' is active
Experiment 'pythonJediLSP' is active
Experiment 'pythonDiscoveryModuleWithoutWatcher' is active
Experiment 'pythonTensorboardExperiment' is active
Experiment 'PythonPyTorchProfiler' is active
Experiment 'pythonDeprecatePythonPath' is active
Experiment 'pythonSortEnvs' is active
Experiment 'pythonRunFailedTestsButtonDisplayed' is active
Experiment 'pythonRefreshTestsButtonDisplayed' is active
Experiment 'pythonRememberDebugConfig' is active
> conda info --json
> /bin/python ~/.vscode/extensions/ms-python.python-2021.11.1422169775/pythonFiles/interpreterInfo.py
> /sbin/python ~/.vscode/extensions/ms-python.python-2021.11.1422169775/pythonFiles/interpreterInfo.py
> /usr/bin/python ~/.vscode/extensions/ms-python.python-2021.11.1422169775/pythonFiles/interpreterInfo.py
> /usr/sbin/python ~/.vscode/extensions/ms-python.python-2021.11.1422169775/pythonFiles/interpreterInfo.py
Python interpreter path: /bin/python
Starting Pylance language server.
> conda --version
> /bin/python -m pylint --max-line-length=120 --enable=W0614,W0611 --disable=E1101,I1101,E0401,C0111,W0212,C0103 --msg-template='{line},{column},{category},{symbol}:{msg} --reports=n --output-format=text ~/path/to/file/models/stock_warehouse.py
cwd: ~/Downloads/path
##########Linting Output - pylint##########


This is the end of the very very long log that was generated after the 'Peek References' command is started in "Python Language Server":
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/safe_eval.py (1ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/tools/test_reports.py [fs read 0ms] (3ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/test_reports.py (1ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/tools/view_validation.py [fs read 0ms] (3ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/view_validation.py (1ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/tools/which.py [fs read 0ms] (1ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/which.py (1ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/tools/win32.py [fs read 0ms] (0ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/win32.py (0ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/tools/json/__init__.py [fs read 0ms] (0ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/json/__init__.py (0ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/tools/json/scriptsafe.py [fs read 1ms] (1ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/json/scriptsafe.py ...
[FG]   parsing: /home/username/.vscode/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/json/__init__.pyi [fs read 0ms] (0ms)
[FG]   binding: /home/username/.vscode/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/json/__init__.pyi (0ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/tools/json/scriptsafe.py (1ms)
[FG] parsing: /home/username/path/to/file/odoo/src/odoo/upgrade/__init__.py [fs read 1ms] (1ms)
[FG] binding: /home/username/path/to/file/odoo/src/odoo/upgrade/__init__.py (0ms)
[FG] parsing: /home/username/path/to/file/odoo/src/setup/odoo-wsgi.example.py [fs read 0ms] (1ms)
[FG] binding: /home/username/path/to/file/odoo/src/setup/odoo-wsgi.example.py (0ms)
[FG] parsing: /home/username/path/to/file/odoo/src/setup/package.py [fs read 0ms] (9ms)
[FG] binding: /home/username/path/to/file/odoo/src/setup/package.py (2ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange


This is how it looks like second time (this is all of it):
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
```

## Code Snippet / Additional information

A large python codebase is needed with lots of files, so generating the tags should take several seconds. This way it`s visible if something not working as expected.

**Comment by judej** [comment]
Thanks for the report. Is the code accessible for us to repro this problem?

**Comment by danieltodor** [comment]
I did some more tests, and recorded the time after the peek editor opens.
The opening time is only dependent on the number of files in the project directory.
And the test also shows that the opening time is not linearly increasing with the number of files.

I mostly work on the project with 53K files, so for me it feels like eternity when i search references.

```txt
Poject size          Number of files          First peek editor          Additional peek editors
   1.8GB                    53K                      17s                           6s
   4.5GB                    29K                      10s                           1.5s
   500MB                    27K                      7.5s                          1s
   150MB                    11K                      4s                            0.1s
```

Our projects are private, but if you download several branches to the same directory from this project, you can get to 1.8GB after 3-4 branches.
https://github.com/odoo/odoo

**Comment by bschnurr** [comment]
You could try adding a pyrightconfig.json in your rootfolder with an "exclude" section; see [microsoft/pyright@main/docs/configuration.md#main-pyright-config-options](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#pyright-configuration) for more info.


Thank might help with perf..

**Comment by erictraut** [comment]
@danieltodor, you mentioned that both "Peek References" and "Peek Definition" were slow. Locating the definition of a symbol should be very fast (almost instantaneous) and should not depend on the number of files. Can you confirm whether the issue is limited to "Peek References" or applies to "Peek Definition" also? I just want to make sure that we're focusing on the right problem.

**Comment by erictraut** [comment]
I was able to repro the times you are seeing. As you suggested, I cloned the odoo repo and made 4x copies. This results in about 17K ".py" files. The "Peek References" took about 19.7s the first time and 5.0s on subsequent searches.

I've implemented a few optimizations that will help with large code bases. With these optimizations in place, the time for the first "Peek References" dropped to 6.8s, and subsequent searches dropped to 3.2s. The optimization involves skipping the parsing of file contents if there is no text match for the reference symbol's text. That means that the speedup depends on whether the symbol's name is likely to be found in a source file. If the symbol name is "a", this optimization won't be as effective as if the symbol is called "a_really_long_and_specific_symbol_name".

I'll also note that the "Peek Reference" feature in VS Code is completely synchronous, whereas the "Find All References" (which performs the same operation but displays the results in the left panel) is asynchronous and displays results incrementally. So you may find that "Find All References" is preferable when used within a project with a large file count.

**Comment by danieltodor** [comment]
> @danieltodor, you mentioned that both "Peek References" and "Peek Definition" were slow. Locating the definition of a symbol should be very fast (almost instantaneous) and should not depend on the number of files. Can you confirm whether the issue is limited to "Peek References" or applies to "Peek Definition" also? I just want to make sure that we're focusing on the right problem.

I experience the same performance with the `Peek Definition` also.

**Comment by danieltodor** [comment]
> I was able to repro the times you are seeing. As you suggested, I cloned the odoo repo and made 4x copies. This results in about 17K ".py" files. The "Peek References" took about 19.7s the first time and 5.0s on subsequent searches.
> 
> I've implemented a few optimizations that will help with large code bases. With these optimizations in place, the time for the first "Peek References" dropped to 6.8s, and subsequent searches dropped to 3.2s. The optimization involves skipping the parsing of file contents if there is no text match for the reference symbol's text. That means that the speedup depends on whether the symbol's name is likely to be found in a source file. If the symbol name is "a", this optimization won't be as effective as if the symbol is called "a_really_long_and_specific_symbol_name".
> 
> I'll also note that the "Peek Reference" feature in VS Code is completely synchronous, whereas the "Find All References" (which performs the same operation but displays the results in the left panel) is asynchronous and displays results incrementally. So you may find that "Find All References" is preferable when used within a project with a large file count.

That was fast. Thank you.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2021.12.0, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2021120-2-december-2021)
